### PR TITLE
Task/make action response obsolete

### PR DESCRIPTION
**Pull request by mhsdesign**
moved to neos https://github.com/neos/flow-development-collection/pull/3294

### CODE REVIEW

File: Neos.Flow/Classes/Mvc/Controller/ActionController.php [hunk 1]
```diff
@@ -847,10 +845,7 @@ protected function renderView(ActionResponse $response): ActionResponse
         }
 
         if ($result instanceof ResponseInterface) {
-            $response->replaceHttpResponse($result);
-            if ($result->hasHeader('Content-Type')) {
-                $response->setContentType($result->getHeaderLine('Content-Type'));
-            }
+            return $result;
         }
```
Comment (mhsdesign) [inline Neos.Flow/Classes/Mvc/Controller/ActionController.php:853]: this is the best reason to do this.

File: Neos.FluidAdaptor/Classes/Core/Widget/AbstractWidgetViewHelper.php [hunk 2]
```diff
@@ -241,17 +241,17 @@ protected function initiateSubRequest()
...
 
                 // We need to make sure to not merge content up into the parent ActionResponse because that _could_ break the parent response.
-                $content = $subResponse->getContent();
-                $subResponse->setContent('');
+                $content = $subResponse->getBody()->getContents();
+                $subResponse = $subResponse->withBody(Utils::streamFor(''));
             } catch (StopActionException $exception) {
                 $subResponse = $exception->response;
-                $subResponse->mergeIntoParentResponse($this->controllerContext->getResponse());
+                $this->controllerContext->getResponse()->replaceHttpResponse($subResponse);
```
Comment (mhsdesign) [inline Neos.FluidAdaptor/Classes/Core/Widget/AbstractWidgetViewHelper.php:248]: this is currently not 100% the same but we can fix this https://github.com/neos/flow-development-collection/issues/2492

**Comment by kitsunet** [comment]
I am personally not totally opposed to this, BUT it was always meant to be a level of separation just like the ActionRequest is. Granted I see way less reason for the response to be that, especially the way it exists here, but I am sure this will be in question.

**Comment by mhsdesign** [comment]
Yes i know. I also thought a bit about this before i wanted to experiment with this.

The way i see it is that the current `ActionResponse` is since flow 7 broken, as it allows to "replace" the http reponse. This behaviour is in many cases currently undefined and behaves unexpectedly: https://github.com/neos/flow-development-collection/issues/2492

But even  before Flow 7 and prs 7, the ActionResponse was at least partially inconsistent and the abstraction broken.
The response had these properties:

- content `StreamInterface`
- redirectUri `UriInterface`
- statusCode `int`
- contentType `string`
- cookies `Cookie[]`
- headers `array<string, array>`

the first 5 are a well fit abstraction, but the `headers` are an escape hatch to the lower level with possibly undeclared behaviour:

What if i actually set cookies AND a `Set-Cookie` header? Will `getCookie` understand this? What if i set a custom `Location` header AND redirect uri, what will win and what will `getHeader` return? And most importantly contentType vs `Content-Type`.

After reading the code, i can say that the "headers" functionality is just an array carried around with no validation or transformation. So that will cause conflicts one would ran into with duplicated headers set on the final result.

Additionally the mutability of the `ActionResponse` is completely unpredictable to work with. We might offer this as a legacy behaviour in the action controller, but we should prevent leaking this mutable action response anywhere far outside.

The `ViewInterface` must support the usecase to render a psr message (see fusion view) and it doesnt really make sense to transform a lower level abstraction to a higher one as this might not always even be possible and features might not be supported.

in our case transforming the `ResponseInterface` to an actual `ActionResponse` is actually possible (without using the `httpRequest` hack) as the action response allows the lower level headers feature already so that _could_ be a way https://github.com/neos/flow-development-collection/pull/3287

but im clearly in favour of working with the pretty immutable `ResponseInterface` directly.

For the `ActionRequest` it seems to be a different case, as we need the additional controller information and subrequest details.

**Comment by kitsunet** [comment]
> For the ActionRequest it seems to be a different case, as we need the additional controller information and subrequest details.

I did actually think about replacing it with `\Psr\Http\Message\ServerRequestInterface::withAttribute` at some point, there is still some stuff to be considered, especially nesting behavior for subrequests (that I think we need for some functionality) but in general we could store our _special_ stuff in custom attributes.

**Comment by mhsdesign** [comment]
@bwaidelich [mentioned](https://github.com/neos/flow-development-collection/issues/2492#issuecomment-857735219) once:

> I remember that this kind of logic holes were the reason why I was in favor of a response abstraction that is not so closely related to the HTTP response.

but im not sure that we can ever found a well fit response abstraction and then we will always still have the problem that a view might return a psr response which is too low level for us to handle, if were not that low level ourselves.

**Comment by mhsdesign** [comment]
Ha it seem this `ActionResponse.setHeader` is not that ooold. It was introduce with Flow 7, and previously one would use a `SetHeader` component:

https://github.com/neos/flow-development-collection/pull/2219

So it seems the abstraction of the response was intact before psr 7.

Now we just need to find out what to do :D

Either remove `httpResponse` and `headers` from the `ActionResponse` or remove the whole `ActionResponse` :D

**Comment by mhsdesign** [comment]
wait ill move the conversation to a flow issue

lets continue our discussion here:

-> https://github.com/neos/flow-development-collection/issues/3289


Pr will also be moved to neos: https://github.com/neos/flow-development-collection/pull/3294

**Pull request closed without merging**